File: bin/gap_check.pl
#!/usr/bin/perl

use DateTime;
use SC6::Cam::General;
use SC6::Cam::Config;
use Getopt::Long;
use Data::Dumper;
use Time::ParseDate;

my $mode = "prod";
my $date;
my $ok_gap = 45;
my $only_last_hour;
my $ONE_HOUR = (60*60);

my $c = new SC6::Cam::Config("/usr/local/cam/conf/config.yml");
our $config = $c->getConfig();
our $debug = $c->getDebug();
my $d;

my $result = GetOptions (  "h|help"  => \&usage,
                        "t|date=s"  => \$date,
                        "D|directory=s"  => \$d,
                        "m|mode=s"  => \$mode,
                        "o|only-last-hour"  => \$only_last_hour,
                        "i|gap=i"  => \$ok_gap,
                        "d|debug+"  => \$debug);

my $now = time();

if ( ! $d ) {
    my $dt;
    if ( $date ) {
        print $date, "\n" if ( $debug );
        my ($seconds, $error) = parsedate($date);
        if ( not $seconds ) {
            print "Error can't parse date: $date : $error\n";
            exit(1);
        }
        else {
            print $seconds, "\n" if ( $debug );
            $dt = DateTime->from_epoch(  epoch => $seconds, time_zone => $config->{'General'}->{'Timezone'} );
        }
    }
    else {
        $dt = DateTime->now(  time_zone => $config->{'General'}->{'Timezone'} );
    }
    
    $d = get_image_dir($dt, "orig", $mode);
}

opendir(my $dh, $d) || die;

my $last_was_big = 0;
my $last = 0;
my @files = readdir $dh;
foreach my $f ( sort @files ) {
    next if ( $f eq "." );
    next if ( $f eq ".." );
    my $n = $f;
    $n =~ s/image(\d+)_orig.jpg/$1/; 
    $last = $n if ( $last == 0 );
    if ( $n - $last <= $ok_gap ) {
        $last_was_big = 0;
    }
    else {
        if ( ! $only_last_hour || $now - $n < $ONE_HOUR ) {
	    my $all_secs = $n - $last;
	    my $mins;
	    my $t_string = "$all_secs seconds";
	    if ( $all_secs >= 60 ) {
	        $mins = ( $all_secs / 60 ) % 60;
		if ( $mins == 1 ) {
		    $min_word = "minute";
		}
		else {
		    $min_word = "minutes";
		}
                $seconds = $all_secs % 60; 
		$t_string = "$mins $min_word $seconds seconds ($all_secs secs)";
	    }
            print "Big gap: ", $t_string, " at ", scalar localtime($n);
            if ( $last_was_big ) {
                print " prior was also big\n";
            }
            else {
                print "\n";
            }
       }
       $last_was_big = 1;
    }
    $last = $n;
}

#    my $new =  $d . "/" . $_;
#    $new =~ s/.jpg$//;
#    $new .= "_orig.jpg";
#    print "mv $d/$_ $new\n";
closedir $dh;

sub usage
{
    print "usage: $0 [-d|--debug] [-t|--date=date] [-i|--gap=interval] [-h|--help] \n";
    print "\t-t|--date                - date of files to check\n";
    print "\t-D|--directory           - directory to check (overrules date based directory)\n";
    print "\t-h|--help                - This message\n";
    print "\t-i|--gap                 - OK gap in images\n";
    print "\t-m|--mode                - set mode (prod is default)\n";
    print "\t-o|--only-last-hour      - only report issues from tha last hour\n";
    print "\t--debug                  - print extra debugging information\n";
    exit(1);

}

File: bin/gap_check.py
#!/usr/local/cam/env/bin/python3

import datetime
import os
import sys
from dateutil.parser import parse
import argparse
import yaml
import logging
import logging.config
import time
from pytz import timezone
import smtplib, ssl

ONE_HOUR = 60 * 60

sys.path.append('/usr/local/cam/lib/pythonlib')
import sc6_bucket_shiz
import sc6_bluecodestate
import sc6_sun
import sc6_image
import sc6_general
import sc6_config

def mail_message(subj, body):

    sender = config['Email']['From']
    receivers = (config['Email']['To'])
    msg = "From: {}\nTo: {}\nSubject: {}\n\n{}".format(sender,
                                                       receivers,
                                                       subj,
                                                       body)
    try:
        smtpObj = smtplib.SMTP('localhost')
        smtpObj.sendmail(sender, receivers, msg)
    except SMTPException:
        print("Error: unable to send email")
        sys.exit()
    else:
        logger.debug("Successfully sent email")

body = ''

parser = argparse.ArgumentParser(description='check gaps in cam images.')
parser.add_argument("--mode", dest="mode", required=False,
                    help="prod / test / maybe others?", default="prod")
parser.add_argument("--debug", dest="debug", default=False,
                    required=False,
                    action='store_true',
                    help="should we spew")
parser.add_argument("--email", dest="email", default=False,
                    required=False,
                    action='store_true',
                    help="should we email")
parser.add_argument("--only-last-hour", dest="only_last_hour",
                    default=False,
                    required=False,
                    action='store_true',
                    help="only report issues from tha last hour")
parser.add_argument("--date", dest="date",
                    default=None,
                    help="date of files to check",
                    required=None)
parser.add_argument("--directory", dest="directory",
                    default=None,
                    help="directory to check (overrules date based directory)",
                    required=False)
parser.add_argument("--gap", dest="ok_gap",
                    type=int,
                    default=45,
                    help="OK gap in images",
                    required=False)
args = parser.parse_args()

email = args.email

cfg = sc6_config.Config(config_file='/usr/local/cam/conf/config.yml',
                        mode=args.mode)
config = cfg.get_config()

with open(config['Logging']['LogConfig'], 'rt') as f:
    lconfig = yaml.safe_load(f.read())
logging.config.dictConfig(lconfig)

# create logger
logger = logging.getLogger('cam_collect_images')
logger.setLevel(logging.DEBUG)

debug = config['Debug'] or args.debug

now = time.time()
directory = args.directory

if directory is None:
    if args.date:
        if debug:
            print(args.date)
        try:
            seconds = parse(args.date).timestamp()
        except ValueError:
            print(f"Error can't parse date: {args.date}")
            sys.exit(1)
        if debug:
            print(seconds)
        dt = datetime.datetime.fromtimestamp(seconds, tz=timezone(config['General']['Timezone']))
    else:
        dt = datetime.datetime.now(tz=timezone(config['General']['Timezone']))
    directory = sc6_general.get_image_dir(dt, "orig", args.mode)

if debug:
    print("directory:", directory)

try:
    files = os.listdir(directory)
except OSError as e:
    print(e)
    sys.exit(1)

last_was_big = False
last = 0
for f in sorted(files):
    if f in (".", ".."):
        continue
#    if debug:
#        print(f)
    n = int(f.replace("image", "").replace("_orig.jpg", ""))
    if last == 0:
        last = n
    if n - last <= args.ok_gap:
        last_was_big = False
    else:
        if not args.only_last_hour or now - n < ONE_HOUR:
            all_secs = n - last
            mins = all_secs // 60
            secs = all_secs % 60
            if mins == 1:
                min_word = "minute"
            else:
                min_word = "minutes"
            t_string = f"{mins} {min_word} {secs} seconds ({all_secs} secs)"
            s = f"Big gap: {t_string} at {datetime.datetime.fromtimestamp(n).strftime('%c')}"
            if last_was_big:
                s += " prior was also big"
            if email:
                s += "\n"
                body += s
            else: 
                print(s)
        last_was_big = True
    last = n

if email:
    if body != '':
        mail_message("SC6 Camera: big gap(s)", body)
